src/System.Text.Json/src/System/Text/Json/JsonUtf8Reader.cs

```diff
@@ -0,0 +1,1274 @@
+            Debug.Assert(data.Length > 0);
+
+            consumed = 0;
+            int i = 0;
+
+            ConsumeNumberResult signResult = ConsumeNegativeSign(ref data, ref i);
+            if (signResult == ConsumeNumberResult.NeedMoreData)
+                return false;
+
+            Debug.Assert(signResult == ConsumeNumberResult.OperationIncomplete);
+
+            byte nextByte = data[i];
+            Debug.Assert(nextByte >= '0' && nextByte <= '9');
+
+            if (nextByte == '0')
+            {
+                ConsumeNumberResult result = ConsumeZero(ref data, ref i);
+                if (result == ConsumeNumberResult.NeedMoreData)
+                    return false;
+                if (result == ConsumeNumberResult.Success)
+                    goto Done;
+
+                nextByte = data[i];
+            }
+            else
+            {
+                i++;
+                ConsumeNumberResult result = ConsumeIntegerDigits(ref data, ref i);
+                if (result == ConsumeNumberResult.NeedMoreData)
+                    return false;
+                if (result == ConsumeNumberResult.Success)
+                    goto Done;
+
+                nextByte = data[i];
+                if (nextByte != '.' && nextByte != 'E' && nextByte != 'e')
+                {
+                    _lineBytePosition += i;
+                    ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedNextDigitComponentNotFound, nextByte);
+                }
+            }
+
+            Debug.Assert(nextByte == '.' || nextByte == 'E' || nextByte == 'e');
+
+            if (nextByte == '.')
+            {
+                i++;
+                ConsumeNumberResult result = ConsumeDecimalDigits(ref data, ref i, nextByte);
+                if (result == ConsumeNumberResult.NeedMoreData)
+                    return false;
+                if (result == ConsumeNumberResult.Success)
+                    goto Done;
+
+                nextByte = data[i];
+                if (nextByte != 'E' && nextByte != 'e')
+                {
+                    _lineBytePosition += i;
+                    ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedNextDigitEValueNotFound, nextByte);
+                }
+            }
+
+            Debug.Assert(nextByte == 'E' || nextByte == 'e');
+            i++;
+
+            signResult = ConsumeSign(ref data, ref i, nextByte);
+            if (signResult == ConsumeNumberResult.NeedMoreData)
+                return false;
+
+            Debug.Assert(signResult == ConsumeNumberResult.OperationIncomplete);
+
+            i++;
+            ConsumeNumberResult resultExponent = ConsumeIntegerDigits(ref data, ref i);
+            if (resultExponent == ConsumeNumberResult.NeedMoreData)
+                return false;
+            if (resultExponent == ConsumeNumberResult.Success)
+                goto Done;
+            if (resultExponent == ConsumeNumberResult.OperationIncomplete)
```

Comment:
I think this `if` should be an assert (since the other two states are handled and you only have three currently defined)